docs/proposals/20191017-control-plane.md

```diff
@@ -0,0 +1,544 @@
+
+// ControlPlane is the Schema for the ControlPlane API
+type ControlPlane struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec   ControlPlaneSpec   `json:"spec,omitempty"`
+    Status ControlPlaneStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// ControlPlaneList contains a list of ControlPlane
+type ControlPlaneList struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ListMeta `json:"metadata,omitempty"`
+    Items           []ControlPlane `json:"items"`
+}
+
+func init() {
+    SchemeBuilder.Register(&ControlPlane{}, &ControlPlaneList{})
+}
+
+type ControlPlaneStatusError string
+
+// A more descriptive kind of error that represents an error condition that
+// should be set in the ControlPlane.Status. The "Reason" field is meant for short,
+// enum-style constants meant to be interpreted by control planes. The "Message"
+// field is meant to be read by humans.
+type ControlPlaneError struct {
+    Reason  ControlPlaneStatusError
+    Message string
+}
+
+func (e *ControlPlaneError) Error() string {
+    return e.Message
+}
+
+const (
+    // InvalidConfigurationClusterError indicates that the cluster
+    // configuration is invalid.
+    InvalidConfigurationClusterError ClusterStatusError = "InvalidConfiguration"
+
+    // UnsupportedChangeClusterError indicates that the cluster
+    // spec has been updated in an unsupported way. That cannot be
+    // reconciled.
+    UnsupportedChangeClusterError ClusterStatusError = "UnsupportedChange"
+
+    // CreateClusterError indicates that an error was encountered
+    // when trying to create the cluster.
+    CreateClusterError ClusterStatusError = "CreateError"
+
+    // UpdateClusterError indicates that an error was encountered
+    // when trying to update the cluster.
+    UpdateClusterError ClusterStatusError = "UpdateError"
+
+    // DeleteClusterError indicates that an error was encountered
+    // when trying to delete the cluster.
+    DeleteClusterError ClusterStatusError = "DeleteError"
+)
+```
+
+With the following validations:
+
+- If `ControlPlane.Spec.KubeadmConfigSpec` does not define external etcd (webhook):
+  - `ControlPlane.Spec.Replicas` in [1, 3, 5, 7]
+- `ControlPlane.Spec.Version != ""` (openapi)
+
+And the following defaulting:
+
+- `ControlPlaneSpec.Replicas: 1` (openapi or webhook)
+
+#### Modifications required to existing API Types
+
+- Add `Cluster.Spec.ControlPlaneRef` defined as:
+
+```go
+    // ControlPlaneRef is an optional reference to a provider-specific resource that holds
+    // the details for provisioning the Control Plane for a Cluster
+    // +optional
+    ControlPlaneRef *corev1.ObjectReference `json:"controlPlaneRef,omitempty"`
+```
+
+- Add `Cluster.Status.ControlPlaneReady` defined as:
+
+```go
+    // ControlPlaneReady defines if the control plane is ready
+    // +optional
+    ControlPlaneReady bool `json:"controlPlaneReady"`
```

Comment:
Kubernetes API conventions [highly recommends](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required) using `// +optional` where `omitempty` is present